openedx_tagging/core/tagging/models.py

```diff
@@ -1,39 +1,428 @@
+            " If the taxonomy field is set, then taxonomy.name takes precedence over this field."
+        ),
+    )
+    _value = case_insensitive_char_field(
+        null=False,
+        max_length=500,
+        help_text=_(
+            "User-facing value used for this tag, stored in case tag is null, e.g if taxonomy is free text, or if it"
+            " becomes null (e.g. if the Tag is deleted)."
+            " If the tag field is set, then tag.value takes precedence over this field."
+        ),
+    )
+
+    @property
+    def name(self) -> str:
+        """
+        Returns this tag's name/label.
+
+        If taxonomy is set, then returns its name.
+        Otherwise, returns the cached _name field.
+        """
+        return self.taxonomy.name if self.taxonomy_id else self._name
+
+    @name.setter
+    def name(self, name: str):
+        """
+        Stores to the _name field.
+        """
+        self._name = name
+
+    @property
+    def value(self) -> str:
+        """
+        Returns this tag's value.
+
+        If tag is set, then returns its value.
+        Otherwise, returns the cached _value field.
+        """
+        return self.tag.value if self.tag_id else self._value
+
+    @value.setter
+    def value(self, value: str):
+        """
+        Stores to the _value field.
+        """
+        self._value = value
+
+    @property
+    def tag_ref(self) -> str:
+        """
+        Returns this tag's reference string.
+
+        If tag is set, then returns its id.
+        Otherwise, returns the cached _value field.
+        """
+        return self.tag.id if self.tag_id else self._value
+
+    @property
+    def is_valid(self) -> bool:
+        """
+        Returns True if this ObjectTag represents a valid taxonomy tag.
+
+        A valid ObjectTag must be linked to a Taxonomy, and be a valid tag in that taxonomy.
+        """
+        return self.taxonomy_id and self.taxonomy.validate_object_tag(self)
+
+    def get_lineage(self) -> Lineage:
+        """
+        Returns the lineage of the current tag as a list of value strings.
+
+        If linked to a Tag, returns its lineage.
+        Otherwise, returns an array containing its value string.
+        """
+        return self.tag.get_lineage() if self.tag_id else [self._value]
+
+    def resync(self) -> bool:
+        """
+        Reconciles the stored ObjectTag properties with any changes made to its associated taxonomy or tag.
+
+        This method is useful to propagate changes to a Taxonomy name or Tag value.
+
+        It's also useful for a set of ObjectTags are imported from an external source prior to when a Taxonomy exists to
+        validate or store its available Tags.
+
+        Returns True if anything was changed, False otherwise.
+        """
+        changed = False
+
+        # Locate a taxonomy matching _name
+        if not self.taxonomy_id:
+            taxonomy = Taxonomy.objects.filter(name=self.name, enabled=True).first()
```

Comment:
Should we have some kind of check here if this object is even allowed to be linked to this taxonomy? What if the taxonomy is for a completely different org than the course content that's tagged?

Reply:
Oo good point.. enabled by [821ce62](https://github.com/openedx/openedx-learning/pull/57/commits/821ce62f8e6f127d97f9148b8bb88e0f29247ca1).